active/0000-custom-validation-decorators.md

```diff
@@ -0,0 +1,242 @@
+
+- Using a locally defined validator
+```bicep
+func myLocalValidator(arg string, extraArg int) object => { kind: 'success' }
+
+@myLocalValidator(10)
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "__bicep",
+      "members": {
+        "myLocalValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            },
+            {
+              "name": "extraArg",
+              "type": "int"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "__bicep",
+        "name": "myLocalValidator",
+        "additionalArguments": [10]
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+- Using an imported validator
+```bicep
+import { anImportedValidator } from 'shared.bicep'
+
+@anImportedValidator()
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "__bicep",
+      "members": {
+        "anImportedValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "__bicep",
+        "name": "anImportedValidator"
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+- Using a wildcard-imported validator
+```bicep
+import * as shared from 'shared.bicep'
+
+@shared.anImportedValidator()
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "_1",
+      "members": {
+        "anImportedValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "_1",
+        "name": "anImportedValidator"
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+## Drawbacks
```

Comment:
I suppose another drawback is that the decorator isn't "evaluated" until deploy time, unlike existing decorators; meaning users don't get validation at compile time for values that are known.

Reply:
This seems like an important limitation. It would be possible evaluate the decorator if a literal value is passed in, but we wouldn't be able to do the same kind of type-based evaluation that we do with `minValue` and the like. This will also be true of at least one other planned decorator, though (`pattern` for checking regex matches)